packages/core/src/server/compilerDevMiddleware.ts

```diff
@@ -15,11 +21,54 @@ const require = createRequire(import.meta.url);
 
 type Options = {
   publicPaths: string[];
-  dev: DevConfig;
+  environments: Record<string, EnvironmentContext>;
+  dev: OriginDevConfig;
   server: ServerConfig;
   compiler: Rspack.Compiler | Rspack.MultiCompiler;
 };
 
+type DevConfig = Omit<OriginDevConfig, 'writeToDisk'> & {
+  writeToDisk?:
+    | boolean
+    | ((filename: string, compilationName?: string) => boolean);
+};
+
+// allow to configure dev.writeToDisk in environments
+const formatDevConfig = (
+  config: OriginDevConfig,
+  environments: Record<string, EnvironmentContext>,
+): DevConfig => {
+  const newDevConfig: DevConfig = { ...config };
+  const defaultWriteToDisk = config.writeToDisk;
+
+  const values = Object.values(environments).map(
+    (e) => e.config.dev.writeToDisk,
+  );
+
+  const isSame =
+    values.every((v) => v === true) || values.every((v) => v === false);
+
+  if (isSame) {
+    newDevConfig.writeToDisk = values[0];
+  } else {
+    newDevConfig.writeToDisk = (filePath, compilationName) => {
+      let writeToDisk = defaultWriteToDisk;
+
+      if (compilationName && environments[compilationName]) {
+        writeToDisk =
+          environments[compilationName].config.dev.writeToDisk ??
+          defaultWriteToDisk;
+      }
+
+      return typeof writeToDisk === 'function'
+        ? writeToDisk(filePath)
+        : writeToDisk!;
+    };
+  }
+
+  return newDevConfig;
+};
+
```

Comment:
How about using `Set` to simplify the function:

```suggestion
const formatDevConfig = (
  config: OriginDevConfig,
  environments: Record<string, EnvironmentContext>,
): DevConfig => {
  const writeToDiskValues = Object.values(environments).map(
    (env) => env.config.dev.writeToDisk,
  );
  if (new Set(writeToDiskValues).size === 1) {
    return config;
  }

  return {
    ...config,
    writeToDisk(filePath: string, compilationName?: string) {
      let { writeToDisk } = config;
      if (compilationName && environments[compilationName]) {
        writeToDisk =
          environments[compilationName].config.dev.writeToDisk ?? writeToDisk;
      }
      return typeof writeToDisk === 'function'
        ? writeToDisk(filePath)
        : writeToDisk!;
    },
  };
};
```